CRM/MembershipExtras/Service/MembershipTypeInstalmentCalculator.php

```diff
@@ -0,0 +1,248 @@
+    }
+    if ($noOfInstalment > 1) {
+      $nextInstalmentDate = $startDate->format('Y-m-d');
+      for ($i = 1; $i < $noOfInstalment; $i++) {
+        $intervalSpec = 'P1M';
+        if ($this->schedule == self::QUARTERLY) {
+          $intervalSpec = 'P4M';
+        }
+        $instalmentDate = new DateTime($nextInstalmentDate);
+        $instalmentDate->add(new DateInterval($intervalSpec));
+        $nextInstalmentDate = $instalmentDate->format('Y-m-d');
+        $followingInstalment = new CRM_MembershipExtras_DTO_ScheduleInstalment();
+        $followingInstalment->setInstalmentDate($instalmentDate);
+        $followingInstalment->setInstalmentAmount($instalmentAmount);
+        array_push($instalments, $followingInstalment);
+      }
+    }
+
+    return $instalments;
+  }
+
+  /**
+   * Gets Membership start date
+   *
+   * @param \DateTime|NULL $startDate
+   * @param \DateTime|NULL $endDate
+   * @param \DateTime|NULL $joinDate
+   * @return mixed
+   */
+  private function getMembershipStartDate(DateTime $startDate = NULL, DateTime $endDate = NULL, DateTime $joinDate = NULL) {
+    $membershipDates = $this->membershipTypeDatesCalculator->getDatesForMembershipType(
+      $this->membershipTypes[0],
+      $startDate,
+      $endDate,
+      $joinDate
+    );
+
+    return $membershipDates['start_date'];
+  }
+
+  /**
+   * Validates the membership types passed in to ensure they meets the criteria for calculating
+   *
+   * @throws CRM_MembershipExtras_Exception_InvalidMembershipTypeInstalmentCalculator
+   */
+  private function validateMembershipTypeForInstalment() {
+    $fixedPeriodStartDays = [];
+    foreach ($this->membershipTypes as $membershipType) {
+      if ($membershipType->period_type == 'fixed') {
+        $fixedPeriodStartDays[] = $membershipType->fixed_period_start_day;
+      }
+      if ($membershipType->duration_unit != 'year' || $membershipType->duration_interval != 1) {
+        throw new InvalidMembershipTypeInstalmentCalculator(InvalidMembershipTypeInstalmentCalculator::ONE_YEAR_DURATION);
+      }
+    }
+
+    $fixedPeriodStartDays = array_unique($fixedPeriodStartDays);
+    if (!empty($fixedPeriodStartDays) && count($fixedPeriodStartDays) != 1) {
+      throw new InvalidMembershipTypeInstalmentCalculator(InvalidMembershipTypeInstalmentCalculator::SAME_PERIOD_START_DAY);
+    }
+  }
+
+  /**
+   * Calculates the remaining days from start date to the end of month date.
+   *
+   * @param DateTime $startDate
+   *
+   * @return int
+   */
+  private function calculateRemainingDaysInStartDateMonth(DateTime $startDate) {
+    $endOfMonthDate = new DateTime($startDate->format('Y-m-t'));
+    $interval = $endOfMonthDate->diff($startDate);
+
+    return (int) $interval->format("%a");
+  }
+
+  /**
+   * Calculates the number of days in the month the start date falls in.
+   *
+   * @param DateTime $startDate
+   *
+   * @return int
+   */
+  private function getNumberOfDaysInStartDateMonth(DateTime $startDate) {
```

Comment:
Same for this method, looks like it isn't used anywhere.

Reply:
@MiyaNoctem as describe above, I left it there for using for calculating fixed period type.